File: 传统的设计模式/代理模式.cpp
#include<iostream>
using namespace std;

class Subject
{
public:
	virtual void sailBook() = 0;
};

class RealSubjectBook :public Subject
{
public:
	void  sailBook()
	{
		cout << "sali book" << endl;
	}
};

class dangdangProxy :public Subject
{
public:
	void sailBook()
	{
		Subject*  subject = new RealSubjectBook;
		discount();
		subject->sailBook();
		discount();
	}

	void discount()
	{
		cout << "discount" << endl;
	}
private:
	Subject* m_subject;
};
int maindangdangProxy()
{
	Subject* subject = new dangdangProxy;
	subject->sailBook();
	delete subject;
	subject = nullptr;
	cout << "hello" << endl;
	system("pause");
	return 0;
}

File: 传统的设计模式/代理模式.ts
export interface   Subject{
    sailBook():void;
}

class   RealSubjectBook implements  Subject{
    sailBook():void{
        console.log("sail book");
    }
}

export  class   dangdangProxy   implements  Subject{
    sailBook():void{
        let subject:Subject=new RealSubjectBook();
        this.disconnect();
        subject.sailBook();
        this.disconnect();
    }

    disconnect():void{
        console.log('disconnect');
    }
}